fix(ngcc): support inline `exports.<name> = <decl>` style exports in UMD files

petebacondarwin opened the pull request:
UMD files export values by assigning them to the `exports` variable.
When evaluating expressions ngcc was failing to cope with inline declarations
like `exports.MyComponent = <decl>`.

Fixes #38947

petebacondarwin committed:
fix(ngcc): map `exports` to the current module in UMD files

UMD files export values by assigning them to an `exports` variable.
When evaluating expressions ngcc was failing to cope with expressions
like `exports.MyComponent`.

This commit fixes the `UmdReflectionHost.getDeclarationOfIdentifier()`
method to map the `exports` variable to the current source file.

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -249,6 +249,31 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+        !ts.isFunctionExpression(exportsSymbol.valueDeclaration.parent)) {
+      // There is an `exports` symbol in the local scope that is not a function parameter.
+      // So this `exports` identifier must be a local variable and does not represent the module.
+      return {node: exportsSymbol.valueDeclaration, viaModule: null, known: null, identity: null};
```
Can we just return `null` here to fallback to regular behavior?

petebacondarwin replied:
No because the regular behaviour is broken. It returns the declaration of the `foo` part of `exports.foo`.

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -249,6 +249,31 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+    const exportsSymbol = this.checker.getSymbolsInScope(id, ts.SymbolFlags.Variable)
+                              .find(symbol => symbol.name === 'exports');
+    if (exportsSymbol !== undefined &&
+        !ts.isFunctionExpression(exportsSymbol.valueDeclaration.parent)) {
```
`exportsSymbol.valueDeclaration` might be `undefined` on a bad day

petebacondarwin replied:
Not in JavaScript, I believe. This would only be undefined if the symbol refers to a "type". And `exports` is an identifier in a UMD file here.

JoostK replied:
Yeah, you're probably safe in this context, given that we were able to find the symbol in scope. There are certainly cases where `valueDeclaration` can be unset even for JavaScript, I believe that the `require` symbol for CJS bundles has an undefined value declaration (because it's ambient)

petebacondarwin replied:
I believe we are indeed safe here https://ts-ast-viewer.com/#code/MYewdgzgLgBAHjAvDATgUwI4FcCW6AUA5ALYgiECUA3AFA1pwAOIKUEAdAGZlLxVA

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -249,6 +249,31 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+    if (!isExportIdentifier(id)) {
+      return null;
+    }
+    debugger;
```
Oops

petebacondarwin replied:
Dang!

petebacondarwin committed:
fix(ngcc): handle aliases in UMD export declarations

Sometimes UMD exports appear in the following form:

```
exports.MyClass = alias1 = alias2 = <<declaration>>
```

Previously the declaration of the export would have been captured
as `alias1 = alias2 = <<declaration>>`, which the `PartialInterpreter`
would have failed on, since it cannot handle assignments.

Now we skip over these aliases capturing only the `<<declaration>>`
expression.

Fixes #38947

petebacondarwin committed:
test(compiler-cli): improve error message if a unit test is bad

The message now also reports the name of the predicate function
that failed.

petebacondarwin committed:
refactor(compiler-cli): remove unnecessary constraint on `isDeclarationReference()`

There is no need to check that the `ref.node` is of any particular type
because immediately after this check the entry is tested to see if it passes
`isClassDeclarationReference()`.

The only difference is that the error that is reported is slightly different
in the case that it is a `ref` but not one of the TS node types.

Previously:

```
`Value at position ${idx} in the NgModule.${arrayName} of ${
                className} is not a reference`
```

now

```
`Value at position ${idx} in the NgModule.${arrayName} of ${
                  className} is not a class`
```

Arguably the previous message was wrong, since this entry IS a reference
but is not a class.

petebacondarwin committed:
refactor(ngcc): remove unused imports

The `isAssignment` and `isAssignmentStatement` are not used in this file.

petebacondarwin committed:
refactor(ngcc): rename `ExportStatement` to `ExportsStatement`

This clarifies that this is specifically about statements of the form
`exports.<name> = <declaration>`, rather than a general export
statement such as `export class <ClassName> { ... }`.

petebacondarwin committed:
refactor(ngcc): simplify and rename `getClassDeclarationFromInnerDeclaration()`

The new function does not try to restrict the kind of AST node that it
finds, leaving that to the caller. This will make it more resuable in the
UMD reflection host.

petebacondarwin committed:
refactor(ngcc): simplify and break up ES2015 functions with helpers

The protected helper functions can then be overridden by subclasses of the
Esm2015ReflectionHost.

petebacondarwin committed:
test(ngcc): use `isNamedFunctionDeclaration()` in UMD tests

This makes these tests more resilient to changes in the test code
structure. For example switching from

```
var SomeClass = <implementation>;
exports.SomeClass = SomeClass;
```

to

```
exports.SomeClass = <implementation>;
```

petebacondarwin committed:
test(compiler-cli): make the `getDeclaration()` utility more resilient to code format

Previously `getDeclaration()` would only return the first node that matched
the name passed in and then assert the predicate on this single node.
It also only considered a subset of possible declaration types that we might
care about.

Now the function will parse the whole tree collecting an array of all the
nodes that match the name. It then filters this array based on the predicate
and only errors if the filtered array is empty.

This makes this function much more resilient to more esoteric code formats
such as UMD.

petebacondarwin committed:
refactor(compiler-cli): implement `DeclarationNode` node type

Previously the `ConcreteDeclaration` and `InlineDeclaration` had
different properties for the underlying node type. And the `InlineDeclaration`
did not store a value that represented its declaration.

It turns out that a natural declaration node for an inline type is the
expression. For example in UMD/CommonJS this would be the `exports.<name>`
property access node.

So this expression is now used for the `node` of `InlineDeclaration` types
and the `expression` property is dropped.

To support this the codebase has been refactored to use a new `DeclarationNode`
type which is a union of `ts.Declaration|ts.Expression` instead of `ts.Declaration`
throughout.

petebacondarwin committed:
refactor(compiler-cli): move map creation to avoid unnecessary work

If the `symbol` for the given `node` does not exist then there is no
point in creating the `map`.

petebacondarwin committed:
test(ngcc): fix incorrect test setup

The `SIMPLE_CLASS_FILE` contained a `ChildClass` that had an
internal aliases implementation and extended a `SuperClass` base
class. The call to `__extends` was using the wrong argument for
the child class.

petebacondarwin committed:
test(ngcc): use `isNamedDeclaration()` helper to simplify tests

Previously these tests were checking multiple specific expression
types. The new helper function is more general and will also support
`PropertyAccessExpression` nodes for `InlineDeclaration` types.

petebacondarwin committed:
fix(ngcc): support inline export declarations in UMD files

Previously, any declarations that were defined "inline" were not
recognised by the `UmdReflectionHost`.

For example, the following syntax was completely unrecognized:

```ts
var Foo_1;
exports.Foo = Foo_1 = (function() {
  function Foo() {}
  return Foo;
})();
exports.Foo = Foo_1 = __decorate(SomeDecorator, Foo);
```

Such inline classes were ignored and not processed by ngcc.

This lack of processing led to failures in Ivy applications that relied
on UMD formats of libraries such as `syncfusion/ej2-angular-ui-components`.

Now all known inline UMD exports are recognized and processed accordingly.

Fixes #38947

alxhub commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -530,10 +530,37 @@ export interface EnumMember {
+ * In these case we just use a `ts.Expression`. In this case the declaration expression would be
+ * `exports.someVar`.
+ */
+export type DeclarationNode = ts.Declaration|ts.Expression;
```
I believe the union of these two is just `ts.Node`.

petebacondarwin replied:
There are lots of `ts.Node` types that cannot be assigned to `ts.Declaration` nor `ts.Expression`:

```
ArrayBindingPattern
Bundle
CaseBlock
CaseClause
CatchClause
ComputedPropertyName
Decorator
DefaultClause
ExternalModuleReference
HeritageClause
InputFiles
JSDoc
JSDocTag
JsxClosingElement
LiteralLikeNode
NamedExports
NamedImports
ObjectBindingPattern
PrivateIdentifier
QualifiedName
Statement
SyntaxList
TemplateSpan
Token<...>
TypeNode
UnparsedSection
UnparsedSource
VariableDeclarationList
```

Despite this, I thought I would see if that made any difference by changing the `DeclarationNode` to `ts.Node`. Initial results indicate that this fails compilation due to a lack of overlap between a number of types.

It is feasible that we could tweak the types in these places to get it to work, but this would probably result in more churn to the code.

alxhub replied:
👍

petebacondarwin committed:
fix(compiler-cli): support namespaced query types in directives

Previously directive "queries" that relied upon a namespaced type

```ts
queries: {
  'mcontent': new core.ContentChild('test2'),
}
```

caused an error to be thrown. This is now supported.

gkalpak commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -93,7 +228,11 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
-        moduleMap.set(exportDeclaration.name, exportDeclaration.declaration);
+        if (!moduleMap.has(exportDeclaration.name)) {
+          // We assume that the first `exports.<name>` is the actual declaration, and that any
+          // subsequent statements that match are decorating the original declaration.
```
Consider adding a code example here.

gkalpak commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -14,53 +14,59 @@ import {MockLogger} from '../../../src/ngtsc/logging/testing';
+      };
+      const expDecl = (name: string, noVar = false) =>
+          mode === 'exported vars' ? `${noVar ? '' : 'var '}${name}` : `exports.${name}`;
+      const varExp = (name: string) => mode === 'exported vars' ? `exports.${name} = ${name};` : '';
```
OOC, what do `expDecl` and `varExp` stand for?

petebacondarwin replied:
`expDecl` -> export declaration : e.g. `exports.<name> = ...` or `var <name> = ...`
`varExp` -> variable export: e.g. `exports.<name> = <name>` (used when the export was defined as a variable.

petebacondarwin replied:
I know its horrible but I couldn't come up with something better that was succinct enough to appear all over the tests as these do.

gkalpak replied:
I find the names confusing, but I can't propose better ones.
Maybe add https://github.com/angular/angular/pull/38959#discussion_r499180679 as a comment above each to make their purpose more clear.

gkalpak reviewed (changes requested):
Basically lgtm (with a few (mostly minor) comments) ✨ 
Nice, clean commits ❤️

gkalpak commented:
Oh, I forgot to mention there are a couple of commit message typos in **refactor(compiler-cli): remove unnecessary constraint on `isDeclarationReference()`**:
- `the entry is check` --> `the entry is checked`
- `Previosly` --> `Previously`

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -595,7 +595,5 @@ function isNgModule(node: ClassDeclaration, compilation: ScopeData): boolean {
 }
 
 function isDeclarationReference(ref: any): ref is Reference<ts.Declaration> {
```
Yes but only if we loosen the type of the `ref` parameter in `isClassDeclarationReference(ref: Reference<ts.Declaration>)` method.  I'll do that.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -44,8 +44,8 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
   getDeclarationOfIdentifier(id: ts.Identifier): Declaration|null {
-    return this.getUmdModuleDeclaration(id) || this.getUmdDeclaration(id) ||
-        super.getDeclarationOfIdentifier(id);
+    return this.getExportsDeclaration(id) || this.getUmdModuleDeclaration(id) ||
```
Yeah but I thought I would leave that as an exercise for the reviewer 😛 

I don't know if CommonJS generally has this problem, since this format is not generated by ng-packagr and probably most CommonJS formats are not bundled using rollup.

So I would like to defer "fixing" CommonJS until it is a problem. Not because it wouldn't benefit, but because there are more pressing tasks.

gkalpak replied:
Agreed :+1:

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -269,9 +270,7 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
-    return {
-      node: id.getSourceFile(), viaModule: null, known: null, identity: null
-    }
+    return {node: id.getSourceFile(), viaModule: null, known: null, identity: null};
```
Moved to the previous commit where the linting error was introduced.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -2581,10 +2582,12 @@ function isTopLevel(node: ts.Node): boolean {
+    return null;
+  }
+
     // It might be the function expression inside the IIFE. We need to go 5 levels up...
```
Yep, it got picked up in a later commit. Will fix.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -310,9 +311,9 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
-    const declaration = outerClassNode !== null ?
-        this.getDeclarationOfIdentifier(outerClassNode.name) :
+    const outerNode = getOuterNodeFromInnerDeclaration(declarationNode);
+    const declaration = outerNode !== null && isNamedVariableDeclaration(outerNode) ?
```
Good point. It turns out that `isNamedVariableDeclaration()` is not checking that `outerNode.name` is actually an identifier. It is just casting it. I will add a new commit to ensure that this and its friends are doing the correct check.

petebacondarwin replied:
New commit added to the end of the PR.

petebacondarwin committed:
refactor(compiler-cli): ensure `isNamed....()` helpers check name is identity

Previously the `node.name` property was only checked to ensure it was
defined. But that meant that it was a `ts.BindingName`, which also includes
`ts.BindingPattern`, which we do not support. But these helper methods were
forcefully casting the value to `ts.Identifier.

Now we also check that the `node.name` is actually an `ts.Identifier`.

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -971,7 +971,7 @@ runInEachFileSystem(() => {
         const host = createHost(bundle, new UmdReflectionHost(new MockLogger(), false, bundle));
         const classNode = getDeclaration(
-            bundle.program, SOME_DIRECTIVE_FILE.name, 'SomeDirective', isNamedVariableDeclaration);
+            bundle.program, SOME_DIRECTIVE_FILE.name, 'SomeDirective', isNamedFunctionDeclaration);
```
Fair enough.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -50,52 +50,64 @@ export function makeProgram(
+ * Search the file specified by `fileName` in the given `program` for a declaration that has the
+ * name `name` and passes the `predicate` function.
+ *
+ * An error will be thrown if there is not at least one AST node with the given `name` and passes
```
Personally I think all three are correct! Which would you prefer?

gkalpak replied:
You're the native speaker :smiley:
I'm happy with whatever you're happy.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -50,52 +50,64 @@ export function makeProgram(
-  if (!assert(chosenDecl)) {
-    throw new Error(`Symbol ${name} from ${fileName} is a ${
-        ts.SyntaxKind[chosenDecl.kind]}. Expected it to pass predicate "${assert.name}()".`);
+  const chosenDecl = chosenDecls.filter(assert)[0];
```
Right. Originally I was checking that exactly one passed the predicate, so I doing something like:

```ts
const filtered = chosenDecls.filter(assert);
if (filtered.length !== 1) { throw ... }
const chosenDecl = filtered[0];
```

But then I found that this was too restrictive. I'll change it to a `find()`. Thanks

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/commonjs_host_spec.ts:
```diff
@@ -2422,9 +2423,8 @@ exports.MissingClass2 = MissingClass2;
           const exportDeclarations = host.getExportsOfModule(file);
           expect(exportDeclarations).not.toBeNull();
           const decl = exportDeclarations!.get('directives') as InlineDeclaration;
```
It does not - but it does then require a `!`. Since it could be `undefined`.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/imports/src/emitter.ts:
```diff
@@ -200,13 +200,9 @@ export class AbsoluteModuleStrategy implements ReferenceEmitStrategy {
-      if (declaration.node !== null) {
-        exportMap.set(declaration.node, name);
-      }
+      exportMap.set(declaration.node, name);
```
It appeared to me that skipping these was no longer needed since inline declarations now have `node` that can be emitted.

gkalpak replied:
But do we want to have inline declarations in `exportMap`? Even if it's harmless having them in there, I assume we don't need them so why not skip them?

petebacondarwin replied:
Absolutely, that is the main point of this PR, no?
This code is referenced when trying to find the declaration of an import from a 3rd party library. If that library is in UMD format and has "inline" exports then skipping them would result in the import failing to be found.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -234,9 +234,7 @@ export class StaticInterpreter {
     // The identifier's declaration is either concrete (a ts.Declaration exists for it) or inline
     // (a direct reference to a ts.Expression).
     // TODO(alxhub): remove cast once TS is upgraded in g3.
```
Thanks. How did I miss this?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -434,14 +434,14 @@ export class NgModuleDecoratorHandler implements
       let typeRef = valueRef;
       let typeNode = this.reflector.getDtsDeclaration(typeRef.node);
-      if (typeNode !== null && ts.isClassDeclaration(typeNode)) {
+      if (typeNode !== null && isNamedClassDeclaration(typeNode)) {
```
Because we need it to have `{name: ts.Identifier}` so that it is copatible with the `typeRef` variable on the next line.

petebacondarwin replied:
I guess we could have used `isNamedDeclaration()` instead...

gkalpak replied:
Either works for me :smiley:

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -286,7 +287,8 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
     }
```
You are right. I think we can remove most of the `xxx.node === null` checks.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -286,7 +287,8 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
     // If the declaration already has traits assigned to it, return as is.
-    if (superDeclaration.known !== null || superDeclaration.identity !== null) {
+    if (superDeclaration.known !== null ||
+        isConcreteDeclaration(superDeclaration) && superDeclaration.identity !== null) {
```
I couldn't work out how to make it redundant. If I remove it then `superDeclaration.identity` is not valid. Can you give me some more guidance here?

gkalpak replied:
This is what I meant:

```diff
   getDeclarationOfIdentifier(id: ts.Identifier): Declaration|null {
     const superDeclaration = super.getDeclarationOfIdentifier(id);

     // If no declaration was found or it's an inline declaration, return as is.
-    if (superDeclaration === null) {
+    if (superDeclaration === null || !isConcreteDeclaration(superDeclaration)) {
       return superDeclaration;
     }

     // If the declaration already has traits assigned to it, return as is.
-    if (superDeclaration.known !== null ||
-        isConcreteDeclaration(superDeclaration) && superDeclaration.identity !== null) {
+    if (superDeclaration.known !== null || superDeclaration.identity !== null) {
       return superDeclaration;
     }

gkalpak replied:
Looking at it again, we probably don't want to exit early if it is an inline declaration any more.
In which case, we just need to update the comment above the first `if` block (and keep the code as is):

```diff
-// If no declaration was found or it's an inline declaration, return as is.
+// If no declaration was found, return null.
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -95,10 +95,10 @@ runInEachFileSystem(() => {
           logs.push(`compile: ${(decl as any).name.text}@${analysis.decoratorName} (resolved: ${
               analysis.resolved})`);
-          return `@${analysis.decoratorName} (compiled)`;
+          return `@${analysis.decoratorName} (compiled)` as any;
```
I removed the `as any` from the earlier line as suggested..

petebacondarwin commented:
I addressed most of your comments @gkalpak except for:

https://github.com/angular/angular/pull/38959#discussion_r499139907
and
https://github.com/angular/angular/pull/38959#discussion_r499162549

ayazhafiz reviewed (approved):
LGTM for language service

gkalpak commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
````diff
@@ -370,3 +563,27 @@ function getRequiredModulePath(wrapperFn: ts.FunctionExpression, paramIndex: num
+ * exports.MyClass = alias1 = alias2 = <<declaration>>
+ * ```
+ *
+ * @node the expression to parse
````
```suggestion
 * @param node the expression to parse
```

petebacondarwin replied:
😱

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1624,7 +1638,7 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
     const imp = this.getImportOfExpression(typeExpression);
     const decl = this.getDeclarationOfExpression(typeExpression);
-    if (imp === null || decl === null || decl.node === null) {
+    if (imp === null || decl === null) {
```
No I don't think so. Previously we couldn't handle inline exports, which is why there are all these bailouts.
But now they are fully supported, and if there is a declaration that is inline but also has an import (`imp`) then it should not be a `TypeValueReferenceKind.LOCAL`.

gkalpak replied:
I'm confused now. Can you give an example of a `typeExpression` that has both an import and an inline declaration.

petebacondarwin replied:
I think this is a scenario: https://github.com/angular/angular/blob/e4424863c2cebdf3786f82c336294195d7733c93/packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts#L580-L606

Here the `typeToValue()` function is needed to ensure that the "type" `ChangeDetectorRef` is also considered a value so that we get the correct information about it as an export.

If the `@angular/core` library in this test was packaged via ng-packagr as UMD then all its exports would be "inline".
If we didn't make the change here then these exports would be skipped, I believe.

gkalpak replied:
I'll take your word for it 😆

gkalpak reviewed (approved):
:rocket:

JoostK commented on packages/compiler-cli/ngcc/src/analysis/ngcc_references_registry.ts:
```diff
@@ -29,12 +29,13 @@ export class NgccReferencesRegistry implements ReferencesRegistry {
       if (ref.bestGuessOwningModule === null && hasNameIdentifier(ref.node)) {
         const declaration = this.host.getDeclarationOfIdentifier(ref.node.name);
-        if (declaration && declaration.node !== null && hasNameIdentifier(declaration.node)) {
+        if (declaration && isConcreteDeclaration(declaration) &&
```
If we're filtering out inline declarations here, wouldn't that mean that such declarations won't be identified as needing a public export (see `PrivateDeclarationsAnalyzer`)?

petebacondarwin replied:
I think you are right.
In any case it will be easier to diagnose issue where we include an inline declaration when we should not have, rather than the other way around.
I'll remove this guard.

JoostK commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -220,6 +207,59 @@ export function isRequireCall(node: ts.Node): node is RequireCall {
+ * `exports.<foo> = <bar>`.
+ */
+export function isExportsAssignment(expr: ts.Node): expr is ExportsAssignment {
+  return ts.isBinaryExpression(expr) && ts.isPropertyAccessExpression(expr.left) &&
```
Do we want to verify the binary expr's operator is an `=` token?

petebacondarwin replied:
Good idea.

petebacondarwin replied:
I found a helper called `isAssignment()`!

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -44,8 +46,48 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+      return superDeclaration;
+    }
+
+    // If the intiailizer is an identifier then this is a reference to some other declaration.
+    // Start the process over with this identifier as the target.
+    if (ts.isIdentifier(outerNode.right)) {
+      return this.getDeclarationOfIdentifier(outerNode.right);
```
Is this really needed? It seems unexpected to me that we unroll this apparently intermediate declaration; the static interpreter should typically do that.

Additionally, do we need to worry about alias declarations here? `exports.Foo = Foo_1 =(function() {...})();`

petebacondarwin replied:
Let me think about this...

petebacondarwin replied:
Looking at the implementation of `getOuterNodeFromInnerDeclaration()` it is not possible for the `outerNode` to be anything other than `SomeIdentifier = <IIFE>`. So it is not possible for the the RHS to be an identifier. Therefore I think you are right that this is redundant.

Regarding aliases, I think that this needs to be fixed in `getOuterNodeFromInnerDeclaration()` to make sure that it returns the far LHS of any assignments.

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -77,6 +119,100 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+    const innerDeclaration = getInnerClassDeclaration(initializer);
+    if (innerDeclaration !== null) {
+      return this.createClassSymbol(declaration.name, innerDeclaration);
+    }
```
Could you add an explicit return for the `undefined` value?

JoostK commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -77,6 +119,100 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
+
+  /**
+   * Analyze the given statement to see if it corresponds with an exports declaration like
+   * `exports.MyClass = MyClass_1 = class MyClass {};`. If so, the declaration of `MyClass_1`
```
As UMD is typically ES5, the `class MyClass {}` syntax would normally be an IIFE, right?

petebacondarwin replied:
My bad.

JoostK commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -2453,782 +2697,801 @@ runInEachFileSystem(() => {
-              ['xtra2', `xtra2 = 'xtra2'`, _('/xtra_module')],
-            ]);
-      });
+        it('should return a the correct declaration for a decorated class', () => {
```
```suggestion
        it('should return the correct declaration for a decorated class', () => {
```

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -463,12 +463,20 @@ export function extractQueriesFromDecorator(
           'Decorator query metadata must be an instance of a query type');
     }
-    const type = reflector.getImportOfIdentifier(queryExpr.expression);
+    let queryType = ts.isPropertyAccessExpression(queryExpr.expression) ?
```
```suggestion
    const queryType = ts.isPropertyAccessExpression(queryExpr.expression) ?
```

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -530,10 +530,37 @@ export interface EnumMember {
+ * declarations, function declarations etc.
+ * But in some cases there is no `ts.Declaration` that can be used for a declaration such
+ * as when they are declared inline as part of an exported expression, e.g. `exports.someVar = 42`.
+ * In these case we just use a `ts.Expression`. In this case the declaration expression would be
```
```suggestion
 * In these cases we just use a `ts.Expression`. In this case the declaration expression would be
```

petebacondarwin replied:
I generally reworded this block as it felt a bit messy.

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -530,10 +530,37 @@ export interface EnumMember {
+export type DeclarationNode = ts.Declaration|ts.Expression;
+
+/**
+ * The type of a Declaration - whether it's node is concrete (ts.Declaration) or inline
```
```suggestion
 * The type of a Declaration - whether its node is concrete (ts.Declaration) or inline
```

JoostK reviewed (commented):
Went over it pretty quickly in one go, so haven't looked at the individual commits (nor their messages)

JoostK reviewed (approved):
🎉

alxhub commented:
[Presubmit](http://test/OCL:335682260:BASE:335682313:1602008807911:caf4f78a)

atscott commented:
[presubmit](http://test/OCL:336311580:BASE:336311625:1602262357949:ab3293e7)

atscott commented:
Note: removing the merge label as this was found to break an internal build (pinged @petebacondarwin directly with details)
